packages/sprotty-protocol/src/actions.ts

```diff
@@ -0,0 +1,500 @@
+    return hasOwnProperty(object, 'action');
+}
+
+/**
+ * An action describes a change to the model declaratively.
+ * It is a plain data structure, and as such transferable between server and client.
+ */
+export interface Action {
+    kind: string;
+}
+
+export function isAction(object?: unknown): object is Action {
+    return hasOwnProperty<string, string>(object, 'kind', 'string');
+}
+
+/**
+ * A request action is tied to the expectation of receiving a corresponding response action.
+ * The `requestId` property is used to match the received response with the original request.
+ */
+export interface RequestAction<Res extends ResponseAction> extends Action {
+    requestId: string
+}
+
+export function isRequestAction(object?: Action): object is RequestAction<ResponseAction> {
+    return hasOwnProperty<string, string>(object, 'requestId', 'string');
+}
+
+let nextRequestId = 1;
+/**
+ * Generate a unique `requestId` for a request action.
+ */
+export function generateRequestId(): string {
+    return (nextRequestId++).toString();
+}
+
+/**
+ * A response action is sent to respond to a request action. The `responseId` must match
+ * the `requestId` of the preceding request. In case the `responseId` is empty or undefined,
+ * the action is handled as standalone, i.e. it was fired without a preceding request.
+ */
+export interface ResponseAction extends Action {
+    responseId: string;
+}
+
+/**
+ * A response action is sent to respond to a request action. The `responseId` must match
+ * the `requestId` of the preceding request. In case the `responseId` is empty or undefined,
+ * the action is handled as standalone, i.e. it was fired without a preceding request.
+ */
+
+export function isResponseAction(object?: Action): object is ResponseAction {
+    return hasOwnProperty<string, string>(object, 'responseId', 'string');
+}
+
+/**
+ * A reject action is fired to indicate that a request must be rejected.
+ */
+export interface RejectAction extends ResponseAction {
+    kind: typeof RejectAction.KIND
+    message: string
+    detail?: JsonAny
+}
+export namespace RejectAction {
+    export const KIND = 'rejectRequest';
+
+    export function create(options: { message: string, detail?: JsonAny, requestId: string }): RejectAction {
+        return {
+            kind: KIND,
+            message: options.message,
+            detail: options.detail,
+            responseId: options.requestId
+        };
+    }
+}
+
+/**
+ * Sent from the client to the model source (e.g. a DiagramServer) in order to request a model. Usually this
+ * is the first message that is sent to the source, so it is also used to initiate the communication.
+ * The response is a SetModelAction or an UpdateModelAction.
+ */
+export interface RequestModelAction extends RequestAction<SetModelAction> {
+    kind: typeof RequestModelAction.KIND
+    options?: { [key: string]: string | number | boolean }
```

Comment:
I agree, but remember that our current Java server implementation uses `Map<String, String>` as type, you you'll need to be careful to use only primitive values with the Java server.

Reply:
Ah that's a fair  point. I overlooked that the JsonMap type allows `JsonAny` as values and not  `JsonPrimitive`. In that case it might be better to keep it as it.

Reply:
Using `JsonMap` as you proposed could be useful for NodeJS backends. I already changed it that way and I'd propose to keep that.